src/accessControl/acp/index.js

```diff
@@ -0,0 +1,305 @@
+    acp.access.read && acp.access.write
+  );
+}
+
+export function getOrCreatePermission(permissions, webId) {
+  const permission = permissions[webId] || {
+    webId,
+  };
+  permission.acp = permissions.acp || {
+    apply: createAcpMap(),
+    access: createAcpMap(),
+  };
+  return permission;
+}
+
+export function getPolicyUrl(resource, policies) {
+  const resourceUrl = getSourceUrl(resource);
+  const policiesUrl = getSourceUrl(policies);
+  const matchingStart = sharedStart(resourceUrl, policiesUrl);
+  const path = resourceUrl.substr(matchingStart.length);
+  return `${getPoliciesContainerUrl(matchingStart) + path}.ttl`;
+}
+
+export function getOrCreatePolicy(dataset, url) {
+  const existingPolicy = getPolicy(dataset, url);
+  if (existingPolicy) {
+    return { policy: existingPolicy, dataset };
+  }
+  const newPolicy = createPolicy(url);
+  const newDataset = setPolicy(dataset, newPolicy);
+  return { policy: newPolicy, dataset: newDataset };
+}
+
+export async function setAgents(policy, webId, access, mode) {
+  const { agents: existingAgents } = getRequiredRuleOnPolicy(policy);
+  const agentIndex = existingAgents.indexOf(webId);
+  let agents;
+  if (access[mode] && agentIndex === -1) {
+    agents = existingAgents.concat(webId);
+  }
+  if (!access[mode] && agentIndex !== -1) {
+    agents = existingAgents
+      .slice(0, agentIndex)
+      .concat(existingAgents.slice(agentIndex + 1));
+  }
+  return setRequiredRuleOnPolicy(policy, { agents });
+}
+
+export default class AcpAccessControlStrategy {
+  #datasetWithAcr;
+
+  #policyUrl;
+
+  #fetch;
+
+  constructor(datasetWithAcr, policies, fetch) {
+    this.#datasetWithAcr = datasetWithAcr;
+    this.#policyUrl = getPolicyUrl(datasetWithAcr, policies);
+    this.#fetch = fetch;
+  }
+
+  async getPolicies(policyUrls) {
+    const policyResources = await Promise.all(
+      policyUrls.map((url) => getSolidDataset(url, { fetch: this.#fetch }))
+    );
+    const policies = policyResources
+      .map((resource) => getPolicyAll(resource))
+      .flat();
+    return policies.map((policy) => ({
+      modes: getAllowModesOnPolicy(policy),
+      agents: getRequiredRuleOnPolicy(policy).agents,
+    }));
+  }
+
+  async getPermissions() {
+    const permissions = {};
+    // assigning Read, Write, and Append
+    const accessControls = getAccessControlAll(this.#datasetWithAcr);
+    const policyUrls = accessControls
+      .map((ac) => getPolicyUrlAll(ac).filter((url) => url === this.#policyUrl))
+      .flat();
+    (await this.getPolicies(policyUrls)).forEach(({ modes, agents }) =>
+      agents.forEach((webId) => {
+        const permission = getOrCreatePermission(permissions, webId);
+        permission.acp.apply = addAcpModes(permission.acp.apply, modes);
+        permissions[webId] = permission;
+      })
+    );
+    // assigning Control
+    const controlPolicyUrls = getControlPolicyUrlAll(
+      this.#datasetWithAcr
+    ).filter((url) => url === this.#policyUrl);
```

Comment:
Will this not simply lead to

```
const controlPolicyUrls = [this.#policyUrl]
```

?

Reply:
As I understand it this will lead to either `[this.#policyUrl]` or `[]`, given whether there is a policy set for control or not.

Reply:
OK. Just noted it because it looked weird to me to create an array with just either one or zero elements, and then `forEach`'ing over that, but I suppose that makes it easier with `this.getPolicies` :)